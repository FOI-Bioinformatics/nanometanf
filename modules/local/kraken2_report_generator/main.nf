process KRAKEN2_REPORT_GENERATOR {
    tag "${meta.id}"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/krakentools:1.2--pyh5e36f6f_0' :
        'quay.io/biocontainers/krakentools:1.2--pyh5e36f6f_0' }"

    input:
    tuple val(meta), path(cumulative_output)
    tuple val(meta), path(batch_reports)
    path  db

    output:
    tuple val(meta), path("*.cumulative.kraken2.report.txt"), emit: report
    tuple val(meta), path("report_stats.json"), emit: stats
    path  "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ''
    """
    #!/usr/bin/env python3

    import json
    import sys
    import subprocess
    from pathlib import Path

    # Get list of batch report files
    report_files = "${batch_reports}".split()
    print(f"Combining {len(report_files)} batch reports for sample ${meta.id}", file=sys.stderr)

    # Use KrakenTools combine_kreports.py to merge batch reports
    # This sums up the read counts across all batches
    cmd = [
        'combine_kreports.py',
        '-r'
    ] + report_files + [
        '-o', '${prefix}.cumulative.kraken2.report.txt',
        '--display-headers'
    ]

    print(f"Running: {' '.join(cmd)}", file=sys.stderr)

    try:
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        if result.stdout:
            print(result.stdout, file=sys.stderr)
        if result.stderr:
            print(result.stderr, file=sys.stderr)
    except subprocess.CalledProcessError as e:
        print(f"ERROR: combine_kreports.py failed with exit code {e.returncode}", file=sys.stderr)
        print(f"STDOUT: {e.stdout}", file=sys.stderr)
        print(f"STDERR: {e.stderr}", file=sys.stderr)
        sys.exit(1)

    # Calculate statistics from the cumulative output
    total_reads = 0
    classified_reads = 0

    # Count lines in cumulative output (one line per read)
    with open('${cumulative_output}') as f:
        for line in f:
            total_reads += 1
            if line.strip() and not line.startswith('U'):
                classified_reads += 1

    # Generate report statistics JSON
    report_stats = {
        'sample_id': '${meta.id}',
        'total_reads': total_reads,
        'classified_reads': classified_reads,
        'classification_rate': classified_reads / total_reads if total_reads > 0 else 0,
        'num_batches': len(report_files),
        'cumulative_report': '${prefix}.cumulative.kraken2.report.txt'
    }

    with open('report_stats.json', 'w') as stats:
        json.dump(report_stats, stats, indent=2)

    print(f"", file=sys.stderr)
    print(f"Cumulative report generated:", file=sys.stderr)
    print(f"  Total reads: {total_reads}", file=sys.stderr)
    print(f"  Classified: {classified_reads} ({report_stats['classification_rate']*100:.1f}%)", file=sys.stderr)
    print(f"  Batches merged: {len(report_files)}", file=sys.stderr)

    # Generate versions.yml
    with open('versions.yml', 'w') as v:
        v.write('"${task.process}":\n')
        v.write(f'    python: {sys.version.split()[0]}\n')
        # Try to get KrakenTools version
        try:
            kt_version = subprocess.run(
                ['combine_kreports.py', '--version'],
                capture_output=True, text=True
            ).stdout.strip() or '1.2'
        except:
            kt_version = '1.2'
        v.write(f'    krakentools: {kt_version}\n')
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.cumulative.kraken2.report.txt
    echo '{"sample_id": "${meta.id}", "total_reads": 0, "classified_reads": 0}' > report_stats.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 --version | sed 's/Python //g')
        krakentools: 1.2
    END_VERSIONS
    """
}
