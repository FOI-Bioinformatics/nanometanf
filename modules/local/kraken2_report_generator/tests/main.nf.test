nextflow_process {
    name "Test Process KRAKEN2_REPORT_GENERATOR"
    script "../main.nf"
    process "KRAKEN2_REPORT_GENERATOR"
    tag "modules"
    tag "modules_local"
    tag "kraken2_report_generator"
    tag "kraken2"
    tag "incremental"
    tag "krakentools"

    test("generate report from 2 batches - stub") {
        options "-stub"

        when {
            process {
                """
                // Mock cumulative output file
                def cumulative = file("sample1.cumulative.kraken2.output.txt")
                cumulative.text = '''C\tread1\t12345\t100\t12345:1
U\tread2\t0\t100\t0:100
C\tread3\t67890\t100\t67890:1
C\tread4\t12345\t100\t12345:1
'''

                input[0] = [
                    [ id:'sample1' ],
                    cumulative
                ]

                // Mock batch report files
                def report0 = file("sample1_batch0.kraken2.report.txt")
                report0.text = '''100.00\t4\t4\tU\t0\tunclassified
  50.00\t2\t0\tR\t1\troot
  50.00\t2\t2\tD\t12345\tBacteria
'''

                def report1 = file("sample1_batch1.kraken2.report.txt")
                report1.text = '''100.00\t2\t0\tU\t0\tunclassified
  50.00\t1\t0\tR\t1\troot
  50.00\t1\t1\tD\t67890\tViruses
'''

                input[1] = [
                    [ id:'sample1' ],
                    [ report0, report1 ]
                ]

                // Mock database directory
                input[2] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/db/kraken2.tar.gz", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("generate report from 3 batches - stub") {
        options "-stub"

        when {
            process {
                """
                def cumulative = file("sample2.cumulative.kraken2.output.txt")
                cumulative.text = '''C\tread1\t12345\t100\t12345:1
C\tread2\t12345\t100\t12345:1
C\tread3\t67890\t100\t67890:1
U\tread4\t0\t100\t0:100
C\tread5\t11111\t100\t11111:1
C\tread6\t12345\t100\t12345:1
'''

                input[0] = [
                    [ id:'sample2' ],
                    cumulative
                ]

                def report0 = file("sample2_batch0.kraken2.report.txt")
                report0.text = '''100.00\t2\t0\tR\t1\troot'''

                def report1 = file("sample2_batch1.kraken2.report.txt")
                report1.text = '''100.00\t2\t0\tR\t1\troot'''

                def report2 = file("sample2_batch2.kraken2.report.txt")
                report2.text = '''100.00\t2\t0\tR\t1\troot'''

                input[1] = [
                    [ id:'sample2' ],
                    [ report0, report1, report2 ]
                ]

                input[2] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/db/kraken2.tar.gz", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.report },
                { assert process.out.stats },
                { assert snapshot(
                    process.out.report,
                    process.out.stats,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("generate report - single batch edge case - stub") {
        options "-stub"

        when {
            process {
                """
                def cumulative = file("sample3.cumulative.kraken2.output.txt")
                cumulative.text = '''C\tread1\t12345\t100\t12345:1
U\tread2\t0\t100\t0:100
'''

                input[0] = [
                    [ id:'sample3' ],
                    cumulative
                ]

                def report0 = file("sample3_batch0.kraken2.report.txt")
                report0.text = '''100.00\t2\t0\tR\t1\troot
  50.00\t1\t1\tD\t12345\tBacteria
  50.00\t1\t1\tU\t0\tunclassified
'''

                input[1] = [
                    [ id:'sample3' ],
                    [ report0 ]
                ]

                input[2] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/db/kraken2.tar.gz", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.report },
                { assert process.out.stats }
            )
        }
    }

    test("verify statistics calculation - stub") {
        options "-stub"

        when {
            process {
                """
                // Create cumulative output with known counts
                def cumulative = file("sample4.cumulative.kraken2.output.txt")
                cumulative.text = '''C\tread1\t12345\t100\t12345:1
C\tread2\t12345\t100\t12345:1
C\tread3\t67890\t100\t67890:1
U\tread4\t0\t100\t0:100
U\tread5\t0\t100\t0:100
C\tread6\t11111\t100\t11111:1
C\tread7\t12345\t100\t12345:1
C\tread8\t67890\t100\t67890:1
U\tread9\t0\t100\t0:100
C\tread10\t12345\t100\t12345:1
'''
                // Expected: 10 total reads, 7 classified, 3 unclassified

                input[0] = [
                    [ id:'sample4' ],
                    cumulative
                ]

                def report0 = file("sample4_batch0.kraken2.report.txt")
                report0.text = '''100.00\t5\t0\tR\t1\troot'''

                def report1 = file("sample4_batch1.kraken2.report.txt")
                report1.text = '''100.00\t5\t0\tR\t1\troot'''

                input[1] = [
                    [ id:'sample4' ],
                    [ report0, report1 ]
                ]

                input[2] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/db/kraken2.tar.gz", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.report },
                { assert process.out.stats }
            )
        }
    }

    test("metadata preservation - stub") {
        options "-stub"

        when {
            process {
                """
                def cumulative = file("special.cumulative.kraken2.output.txt")
                cumulative.text = '''C\tread1\t12345\t100\t12345:1\n'''

                input[0] = [
                    [ id:'special_sample', single_end:true, sequencer:'MinION' ],
                    cumulative
                ]

                def report0 = file("special_batch0.kraken2.report.txt")
                report0.text = '''100.00\t1\t0\tR\t1\troot'''

                input[1] = [
                    [ id:'special_sample', single_end:true, sequencer:'MinION' ],
                    [ report0 ]
                ]

                input[2] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/db/kraken2.tar.gz", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.report.get(0).get(0).id == 'special_sample' },
                { assert process.out.report.get(0).get(0).single_end == true },
                { assert process.out.report.get(0).get(0).sequencer == 'MinION' }
            )
        }
    }

    test("large scale - 10 batches - stub") {
        options "-stub"

        when {
            process {
                """
                // Generate cumulative output with 100 reads
                def lines = []
                (1..100).each { i ->
                    def taxid = (i % 3 == 0) ? "0" : "12345"
                    def type = (i % 3 == 0) ? "U" : "C"
                    lines << "\${type}\\tread\${i}\\t\${taxid}\\t100\\t\${taxid}:1"
                }

                def cumulative = file("sample_large.cumulative.kraken2.output.txt")
                cumulative.text = lines.join("\\n")

                input[0] = [
                    [ id:'sample_large' ],
                    cumulative
                ]

                // Generate 10 batch reports
                def reports = []
                (0..9).each { i ->
                    def report = file("sample_large_batch\${i}.kraken2.report.txt")
                    report.text = '''100.00\t10\t0\tR\t1\troot'''
                    reports << report
                }

                input[1] = [
                    [ id:'sample_large' ],
                    reports
                ]

                input[2] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/db/kraken2.tar.gz", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.report },
                { assert process.out.stats }
            )
        }
    }
}
