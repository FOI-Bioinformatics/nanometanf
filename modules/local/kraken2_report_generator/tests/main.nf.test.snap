{
    "generate report from 3 batches - stub": {
        "content": [
            [
                [
                    {
                        "id": "sample2"
                    },
                    "sample2.cumulative.kraken2.report.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "sample2"
                    },
                    "report_stats.json:md5,e20644badb75f7c606755560216b6f9f"
                ]
            ],
            [
                "versions.yml:md5,031df40bdbc5b57992a15927efaac92c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T16:57:07.818879"
    },
    "generate report from 2 batches - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "sample1"
                        },
                        "sample1.cumulative.kraken2.report.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "sample1"
                        },
                        "report_stats.json:md5,018398f17757797c43306de66b2721fd"
                    ]
                ],
                "2": [
                    "versions.yml:md5,031df40bdbc5b57992a15927efaac92c"
                ],
                "report": [
                    [
                        {
                            "id": "sample1"
                        },
                        "sample1.cumulative.kraken2.report.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "sample1"
                        },
                        "report_stats.json:md5,018398f17757797c43306de66b2721fd"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,031df40bdbc5b57992a15927efaac92c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-20T16:57:02.699288"
    }
}